### PR TITLE
 👀

### PR DESCRIPTION
**Pull request by d3nwah**

### CODE REVIEW

File: Source/AliveLibAO/Abe.cpp [hunk 1]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+            field_114_gnFrame = gnFrameCount_507670 + 90;
+            field_FC_current_motion = eAbeStates::State_150_Chant_42FD50;
+            field_110_state = 0;
+            SND_SEQ_PlaySeq_4775A0(11u, 0, 1);
+        }
+        return;
+    }
+    if (sInputObject_5009E8.isPressed(dword_4C65B8 | dword_4C65DC))
+    {
+        auto held = sInputObject_5009E8.field_0_pads[sCurrentControllerIndex_5076B8].field_6_held;
```
Comment (mouzedrift) [inline Source/AliveLibAO/Abe.cpp:2804]: make ``held`` const?

File: Source/AliveLibAO/Abe.cpp [hunk 2]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+    if (sInputObject_5009E8.isPressed(sInputKey_Hop_4C65A0))
+    {
+        if (sInputObject_5009E8.isPressed(sInputKey_Up_4C6598))
+        {
+            TryHoist_423420();
+        }
+        else if (!sInputObject_5009E8.isPressed(dword_4C65B8 | dword_4C65DC))
+        {
+            field_FC_current_motion = eAbeStates::State_29_HopBegin_4267B0;
+            auto pObj = IntoBirdPortal_402350(2);
```
Comment (mouzedrift) [inline Source/AliveLibAO/Abe.cpp:2821]: dito

File: Source/AliveLibAO/Abe.cpp [hunk 3]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+            if (pObj)
+            {
+                if (pObj->field_10_portal_type != PortalType::eAbe_0)
+                {
+                    pObj = 0;
```
Comment (mouzedrift) [inline Source/AliveLibAO/Abe.cpp:2826]: nullptr?

File: Source/AliveLibAO/Abe.cpp [hunk 4]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+                field_FE_next_state = eAbeStates::State_133_LiftGrabBegin_42EF20;
+            }
+            else
+            {
+                field_FC_current_motion = eAbeStates::State_133_LiftGrabBegin_42EF20;
+            }
+            return; 
+        }
+
+        auto pHoist = static_cast<Path_Hoist*>(gMap_507BA8.TLV_Get_At_446260(
```
Comment (mouzedrift) [inline Source/AliveLibAO/Abe.cpp:2860]: const?

File: Source/AliveLibAE/Abe.cpp [hunk 5]
```diff
@@ -3175,7 +3175,11 @@ void Abe::State_0_Idle_44EEB0()
 
             case TlvTypes::GrenadeMachine_59:
             {
-                BoomMachine* pMachineButton = static_cast<BoomMachine*>(FindObjectOfType_425180(Types::eGrenadeMachine_66, field_B8_xpos, field_BC_ypos - (field_CC_sprite_scale * FP_FromInteger(25))));
+                BoomMachine* pMachineButton = static_cast<BoomMachine*>(FindObjectOfType_425180(
```
Comment (paulsapps) [inline Source/AliveLibAE/Abe.cpp:3178]: you used auto in the others but not here ;)

File: Source/AliveLibAE/Abe.cpp [hunk 6]
```diff
@@ -3267,49 +3271,55 @@ void Abe::State_0_Idle_44EEB0()
...
-                    const FP xOffSet = ((field_20_animation.field_4_flags.Get(AnimFlags::eBit5_FlipX)) ? FP_FromInteger(15) : FP_FromInteger(-15)) * field_CC_sprite_scale;
-
-                    pThrowable->ctor_431CB0(
-                        field_B8_xpos + xOffSet,
-                        field_BC_ypos + (field_CC_sprite_scale * FP_FromInteger(-50)),
-                        field_20_animation.field_C_render_layer,
-                        field_20_animation.field_14_scale,
-                        field_1A2_throwable_count,
-                        TRUE);
+                    ThrowableTotalIndicator* pThrowable = ae_new<ThrowableTotalIndicator>();
```
Comment (paulsapps) [inline Source/AliveLibAE/Abe.cpp:3286]: ditto for auto?

File: Source/AliveLibAE/Abe.cpp [hunk 7]
```diff
@@ -3267,49 +3271,55 @@ void Abe::State_0_Idle_44EEB0()
...
+            else
             {
-                field_1A2_throwable_count--;
+                field_106_current_motion = eAbeStates::State_34_DunnoBegin_44ECF0;
             }
         }
         else
         {
-            field_106_current_motion = eAbeStates::State_34_DunnoBegin_44ECF0;
+            if (held & sInputKey_DoAction_5550E4) // not throwing, maybe pressing up and pressing action, so do action
```
Comment (paulsapps) [inline Source/AliveLibAE/Abe.cpp:3317]: sInputObject_5BD4E0.IsHeld ?

File: Source/AliveLibAO/Abe.cpp [hunk 8]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+{
+    NOT_IMPLEMENTED();
+    return 0;
+}
+
+static bool IsSameScaleAsHoist(Path_Hoist* pHoist, BaseAliveGameObject* pObj)
+{
+    auto width = pHoist->field_14_bottom_right.field_0_x - pHoist->field_10_top_left.field_0_x;
+
+    if (pObj->field_BC_sprite_scale == FP_FromInteger(1))
```
Comment (paulsapps) [inline Source/AliveLibAO/Abe.cpp:2758]: I  think this is worth a hack comment saying the width is used to determine the layer as there is no layer property in the TLV

File: Source/AliveLibAO/Abe.cpp [hunk 9]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+    else if (pHoist->field_1A_edge_type == Path_Hoist::EdgeType::eRight && pObj->field_10_anim.field_4_flags.Get(AnimFlags::eBit5_FlipX))
+    {
+        return false;
+    }
+    return true;
+}
+
+void Abe::State_0_Idle_423520()
+{
+    Path_TLV *pTlv = nullptr;
```
Comment (paulsapps) [inline Source/AliveLibAO/Abe.cpp:2783]: can be declared more locally to the scope its used?

File: Source/AliveLibAO/Abe.cpp [hunk 10]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+                    }
+                    field_2A8_flags.Clear(Flags_2A8::e2A8_Bit4_Fall_To_Well);
+                    field_F0_pTlv = pTlv;
+                    field_FC_current_motion = eAbeStates::State_77_WellBegin_430F10;
+                    return;
+                }
+                case TlvTypes::WellExpress_45:
+                {
+                    auto well = static_cast<Path_Well_Express*>(pTlv);
+                    if ((well->field_18_scale || field_BC_sprite_scale != FP_FromInteger(1)) &&
```
Comment (paulsapps) [inline Source/AliveLibAO/Abe.cpp:2974]: is AE like this? I think its inverted in AE, checking for == 0 and == 1 is much nicer also

File: Source/AliveLibAO/Abe.cpp [hunk 11]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+                        {
+                            field_FC_current_motion = eAbeStates::State_36_DunnoBegin_423260;
+                        }
+                    }
+                    return;
+                }
+                case TlvTypes::LocalWell_11:
+                {
+                    auto well = static_cast<Path_Well_Local*>(pTlv);
+                    if ((well->field_18_scale || field_BC_sprite_scale != FP_FromInteger(1)) &&
```
Comment (paulsapps) [inline Source/AliveLibAO/Abe.cpp:2961]: same as the other comment, needs inverting to match AE I think

File: Source/AliveLibAO/Abe.cpp [hunk 12]
```diff
@@ -2738,11 +2740,393 @@ void Abe::VOn_Tlv_Collision_421130(Path_TLV *pTlv)
...
+                }
+                if (ResourceManager::GetLoadedResource_4554F0(ResourceManager::Resource_Animation, 32, 0, 0))
+                {
+                    loaded |= 2;
+                }
+                if (ResourceManager::GetLoadedResource_4554F0(ResourceManager::Resource_Animation, 33, 0, 0))
+                {
+                    loaded |= 4;
+                }
+                short rnd = Math_RandomRange_450F20(0, 2);
```
Comment (paulsapps) [inline Source/AliveLibAO/Abe.cpp:3101]: if you're gonna const the other stuff might as well const this as well, also this logic is OddBrain

File: Source/AliveLibAE/Abe.cpp (hunk 5 above)
Comment (d3nwah) [reply Source/AliveLibAE/Abe.cpp:3178]: maybe because I didn't do this line, and just placed endl in it :/

File: Source/AliveLibAE/Abe.cpp (hunk 6 above)
Comment (d3nwah) [reply Source/AliveLibAE/Abe.cpp:3286]: Again, not my doing, just so we're in the clear :tape:

File: Source/AliveLibAO/Abe.cpp (hunk 2 above)
Comment (d3nwah) [reply Source/AliveLibAO/Abe.cpp:2821]: It cannot be const, it gets changed potentially.

File: Source/AliveLibAO/Abe.cpp (hunk 11 above)
Comment (d3nwah) [reply Source/AliveLibAO/Abe.cpp:2961]: It doesn't, same as in AE.

File: Source/AliveLibAO/Abe.cpp (hunk 10 above)
Comment (d3nwah) [reply Source/AliveLibAO/Abe.cpp:2974]: Yes, AE is like this.

File: Source/AliveLibAO/Abe.cpp (hunk 8 above)
Comment (d3nwah) [reply Source/AliveLibAO/Abe.cpp:2759]: added

**Pull request merged**